### PR TITLE
fix kubectl apply error message

### LINKED ISSUE #197: kubectl apply includes raw and useless data into the error message
**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):

"error when retrieving current configuration"

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
/kind bug

**Kubernetes version** (use `kubectl version`):
```
Client Version: version.Info{Major:"", Minor:"", GitVersion:"v1.9.0-beta1", GitCommit:"3258431", GitTreeState:"clean", BuildDate:"2018-01-04T16:20:36Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"10+", GitVersion:"v1.10.0-alpha.1.556+af78b9bac4d34e-dirty", GitCommit:"af78b9bac4d34eeec4adcf48f90845d5b325e081", GitTreeState:"dirty", BuildDate:"2018-01-08T14:52:54Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
```

**What happened**:
`kubectl apply` includes Golang struct into the error message that makes it unreadable and unhelpful to user.

**What you expected to happen**:
1) the error message should be more clear about what's wrong and how I can fix it
2) the error message shouldn't show addresses in the memory that are useless

**How to reproduce it** (as minimally and precisely as possible):
```
$ kubectl apply -f nnp-test.yaml 
error: error when retrieving current configuration of:
&{0xc420094300 0xc420bc6310 default  nnp-test.yaml 0xc4220eb4a8  false}
from server for: "nnp-test.yaml": resource name may not be empty
$ cat nnp-test.yaml
kind: Pod
apiVersion: v1
metadata:
  generateName: testing-nnp-
spec:
  restartPolicy: Never
  initContainers:
  - name: init-container
	image: busybox
	command:
	- grep NoNewPriv /proc/self/status
  containers:
  - name: simple-container
	image: busybox
	command:
	- grep NoNewPriv /proc/self/status
```

**Anything else we need to know**:
```
$ kubectl apply --loglevel=8 -f nnp-test.yaml 
...
I0108 16:42:00.839592    5121 decoder.go:224] decoding stream as YAML
I0108 16:42:00.843520    5121 request.go:624] Error in request: resource name may not be empty
F0108 16:42:00.844222    5121 helpers.go:119] error: error when retrieving current configuration of:
&{0xc4212beb40 0xc420b68070 default  nnp-test.yaml 0xc422ea58c8  false}
from server for: "nnp-test.yaml": resource name may not be empty
```

### PR DESCRIPTION
**Pull request by Lion-Wei**
**What this PR does / why we need it**:
Fix messy code in kubectl apply error message. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes kubernetes/kubectl#197

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```
/sig cli

### CODE REVIEW

**Comment by Lion-Wei** [comment]
/cc @mengqiy 
@php-coder

File: pkg/kubectl/resource/visitor.go [hunk 1]
```diff
@@ -154,6 +154,11 @@ func (i *Info) Refresh(obj runtime.Object, ignoreError bool) error {
 	return nil
 }
 
+// String returns the general purpose string representation
+func (i *Info) String() string {
+	return fmt.Sprintf("namespace: %q, name: %q", i.Namespace, i.Name)
```
Comment (liggitt) [inline pkg/kubectl/resource/visitor.go]: This omits much of the useful info, like the resource and contained object

Comment (Lion-Wei) [reply pkg/kubectl/resource/visitor.go]: Sorry for didn't respond promptly, I changed to cover more information, including `Resource\Version\Object`, now the hint would be like: 
```shell
# ./kubectl apply -f apply.yml 
error: error when retrieving current configuration of:
Resource: "pods", Version: "/v1, Kind=Pod"
Name: "", Namespace: "default"
Object: &{map["metadata":map["generateName":"testing-nnp-" "namespace":"default" "annotations":map["kubectl.kubernetes.io/last-applied-configuration":""]] "spec":map["containers":[map["command":["grep NoNewPriv /proc/self/status"] "image":"busybox" "name":"simple-container"]] "initContainers":[map["command":["grep NoNewPriv /proc/self/status"] "image":"busybox" "name":"init-container"]] "restartPolicy":"Never"] "apiVersion":"v1" "kind":"Pod"]}
from server for: "apply.yml": resource name may not be empty
```

**Comment by Lion-Wei** [comment]
@liggitt I think this is ready for test, can you help me "ok-to -test"? Thanks.

**Comment by Lion-Wei** [comment]
/assign @mengqiy 
Kindly ping,  : )

File: pkg/kubectl/resource/visitor.go [hunk 2]
```diff
@@ -154,6 +154,17 @@ func (i *Info) Refresh(obj runtime.Object, ignoreError bool) error {
 	return nil
 }
 
+// String returns the general purpose string representation
+func (i *Info) String() string {
+	basicInfo := fmt.Sprintf("Name: %q, Namespace: %q\nObject: %+q", i.Name, i.Namespace, i.Object)
```
Comment (mengqiy) [inline pkg/kubectl/resource/visitor.go]: > Object: %+q

Can you paste a example output here?
Do you think using `%+q` will be better than `%#v`?

Comment (Lion-Wei) [reply pkg/kubectl/resource/visitor.go]: Here is the example output of `%+q`
```shell

F0222 12:18:31.699709   77239 helpers.go:119] error: error when retrieving current configuration of:
Resource: "pods", Version: "/v1, Kind=Pod"
Name: "", Namespace: "default"
Object: &{map["apiVersion":"v1" "kind":"Pod" "metadata":map["generateName":"testing-nnp-" "namespace":"default" "annotations":map["kubectl.kubernetes.io/last-applied-configuration":""]] "spec":map["restartPolicy":"Never" "containers":[map["command":["grep NoNewPriv /proc/self/status"] "image":"busybox" "name":"simple-container"]] "initContainers":[map["command":["grep NoNewPriv /proc/self/status"] "image":"busybox" "name":"init-container"]]]]}
from server for: "lw.yml": resource name may not be empty
```
And output of `%#v`
```shell

F0222 12:21:28.568327   81874 helpers.go:119] error: error when retrieving current configuration of:
Resource: "pods", Version: "/v1, Kind=Pod"
Name: "", Namespace: "default"
Object: &unstructured.Unstructured{Object:map[string]interface {}{"apiVersion":"v1", "kind":"Pod", "metadata":map[string]interface {}{"generateName":"testing-nnp-", "namespace":"default", "annotations":map[string]interface {}{"kubectl.kubernetes.io/last-applied-configuration":""}}, "spec":map[string]interface {}{"containers":[]interface {}{map[string]interface {}{"command":[]interface {}{"grep NoNewPriv /proc/self/status"}, "image":"busybox", "name":"simple-container"}}, "initContainers":[]interface {}{map[string]interface {}{"command":[]interface {}{"grep NoNewPriv /proc/self/status"}, "image":"busybox", "name":"init-container"}}, "restartPolicy":"Never"}}}
```

Comment (mengqiy) [reply pkg/kubectl/resource/visitor.go]: >
```
+	always print a sign for numeric values;
	guarantee ASCII-only output for %q (%+q)
```

I think `%+q` is fine.

File: pkg/kubectl/resource/visitor.go [hunk 3]
```diff
@@ -154,6 +154,17 @@ func (i *Info) Refresh(obj runtime.Object, ignoreError bool) error {
...
+	if i.Mapping != nil {
+		mappingInfo := fmt.Sprintf("Resource: %q, Version: %q", i.Mapping.Resource,
```
Comment (mengqiy) [inline pkg/kubectl/resource/visitor.go]: Nit: I think it's more accurate to use `GroupVersionKind` instead of `Version`.

**Review by mengqiy — COMMENTED** [review]
Other LGTM

File: pkg/kubectl/resource/visitor.go (hunk 2 above)
Comment (Lion-Wei) [reply pkg/kubectl/resource/visitor.go]: ：)

File: pkg/kubectl/resource/visitor.go (hunk 3 above)
Comment (Lion-Wei) [reply pkg/kubectl/resource/visitor.go]: Thanks, fixed.

**Comment by Lion-Wei** [comment]
@mengqiy I think it's ready for review and approve. : )

**Comment by Lion-Wei** [comment]
@mengqiy Thanks.

**Pull request merged**